### PR TITLE
[wsu] Run the OVN hybrid overlay

### PR DESCRIPTION
**Pull request by sebsoto**
This commit results in the WSU downloading and running the OVN hybrid
overlay executable. In addition the commit includes a test to ensure
that the overlay was able to run properly, by checking to make sure the
proper annotation has been added to the node.

I have also edited the hack script and dockerfile so that we can edit
the OpenShift network operator in our tests to include a hybrid overlay
configuration.

### CODE REVIEW

File: tools/ansible/tasks/wsu/main.yaml [hunk 1]
```diff
@@ -99,3 +104,25 @@
       delegate_to: localhost
       shell: "oc adm certificate approve {{ item }}"
       with_items: "{{ node_csrs.stdout_lines }}"
+
+    - name: Get bootstrapped node name
+      delegate_to: localhost
+      shell: "oc get no -o wide |awk '/{{ inventory_hostname }}/ {print $1}'"
```
Comment (suhanime) [inline tools/ansible/tasks/wsu/main.yaml]: Nit: Using `oc get nodes` makes context clearer compared to `oc get no`
Nit: Maybe a comment about `{{ inventory_hostname }}` pointing to node ip provided in hostfile?

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 2]
```diff
@@ -37,7 +37,7 @@ var (
 	// The CI-operator uses AWS region `us-east-1` which has the corresponding image ID: ami-0b8d82dea356226d3 for
 	// Microsoft Windows Server 2019 Base with Containers.
 	imageID      = "ami-0b8d82dea356226d3"
-	instanceType = "m4.large"
+	instanceType = "m5a.large"
```
Comment (suhanime) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: We might want to preserve the knowledge somewhere, why we are using AMD based instance on AWS

File: hack/run-wsu-ci-e2e-test.sh [hunk 3]
```diff
@@ -20,8 +20,11 @@ fi
 # The WSU playbook requires the cluster address, we parse that here using oc
 CLUSTER_ADDR=$(oc cluster-info | head -n1 | sed 's/.*\/\/api.//g'| sed 's/:.*//g')
 
+# Set up hybrid networking on the cluster, a requirement of OVNKubernetes on Windows
+oc get Network.operator.openshift.io cluster -o json | jq '.spec.defaultNetwork.ovnKubernetesConfig = {"hybridOverlayConfig":{"hybridClusterNetwork":[{"cidr":"10.132.0.0/14","hostPrefix":23}]}}' | oc apply -f -
```
Comment (aravindhp) [inline hack/run-wsu-ci-e2e-test.sh]: You can do `oc patch network.operator cluster --type=merge -p '{"spec":{"defaultNetwork":{"ovnKubernetesConfig":{"hybridOverlayConfig":{"hybridClusterNetwork":[{"cidr":"10.132.0.0/14","hostPrefix":23}]}}}}}'` instead.

Ideally this and the `Dockerfile.tool` change should be a separate PR. However I am willing to let it slide, if you make these changes a separate commit.

File: hack/run-wsu-ci-e2e-test.sh [hunk 4]
```diff
@@ -20,8 +20,11 @@ fi
...
+
 # Run the test suite
 cd $TEST_DIR
-GO_BUILD_ARGS=CGO_ENABLED=0 GO111MODULE=on CLUSTER_ADDR=$CLUSTER_ADDR WSU_PATH=$WMCO_ROOT/tools/ansible/tasks/wsu/main.yaml go test -v -timeout 20m .
+GO_BUILD_ARGS=CGO_ENABLED=0 GO111MODULE=on CLUSTER_ADDR=$CLUSTER_ADDR WSU_PATH=$WMCO_ROOT/tools/ansible/tasks/wsu/main.yaml go test -v -timeout 30m .
```
Comment (aravindhp) [inline hack/run-wsu-ci-e2e-test.sh]: I see this being increased in every PR. Please say why you are doing this in your commit message.

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 5]
```diff
@@ -36,8 +36,9 @@ var (
 
 	// The CI-operator uses AWS region `us-east-1` which has the corresponding image ID: ami-0b8d82dea356226d3 for
 	// Microsoft Windows Server 2019 Base with Containers.
-	imageID      = "ami-0b8d82dea356226d3"
-	instanceType = "m4.large"
+	imageID = "ami-0b8d82dea356226d3"
+	// Using this instance type, as the Windows hybrid overlay currently does not work on intel processors
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: It has nothing to do with the process but a bug in the Intel 82599 network driver. Please state that here.

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 6]
```diff
@@ -254,3 +260,13 @@ func testWorkerLabelsArePresent(t *testing.T) {
...
+
+// testHybridOverlayAnnotations tests that the correct annotations have been added to the bootstrapped node
+func testHybridOverlayAnnotations(t *testing.T) {
+	windowsNodes, err := k8sclientset.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: windowsLabel})
+	require.NoError(t, err, "Could not get list of Windows nodes")
+	assert.Equalf(t, len(windowsNodes.Items), 1, "expected one windows node to be present but found %v",
+		len(windowsNodes.Items))
+	assert.Contains(t, windowsNodes.Items[0].Annotations, hybridOverlaySubnet)
+	assert.Contains(t, windowsNodes.Items[0].Annotations, hybridOverlayMac)
+}
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: Ideally we should also be testing if the `BaseOpenShiftNetwork` and `OpenShiftNetwork` HNS networks were created on the node. Not sure if it is easy to do that without running `Get-HnsNetwork` remotely.

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go (hunk 5 above)
Comment (aravindhp) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: This needs to be a separate PR as I want this clearly called out and not buried inside a PR that has nothing to do with this change.

File: tools/ansible/tasks/wsu/main.yaml [hunk 7]
```diff
@@ -126,3 +137,17 @@
...
+
+    - name: Start the hybrid overlay
+      win_shell: "Start-Process -NoNewWindow -FilePath \"{{  win_temp_dir.path }}\\hybrid-overlay.exe\" -ArgumentList \"--node  {{ node_name.stdout }} --k8s-kubeconfig c:\\k\\kubeconfig\""
+      async: 60
+      poll: 0
+      register: async_results
+
+    - name: Wait for the hybrid overlay
+      delegate_to: localhost
+      shell: "oc get no {{ node_name.stdout }} -o=jsonpath='{.metadata.annotations}'"
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/main.yaml]: `oc get nodes` please.

File: hack/run-wsu-ci-e2e-test.sh (hunk 3 above)
Comment (sebsoto) [reply hack/run-wsu-ci-e2e-test.sh]: Moved to https://github.com/openshift/windows-machine-config-operator/pull/107

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go (hunk 5 above)
Comment (sebsoto) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: Moved to https://github.com/openshift/windows-machine-config-operator/pull/106
